keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1525 @@
+unnecessary over-allocation.
+
+### Notes/Constraints/Caveats
+
+1. cgroupv1 has been moved to maintenance mode since Kubernetes version 1.31.
+   Hence this feature will be supported only on cgroupv2.
+2. Pod Level Resource Specifications is an **opt-in** feature, and will have no
+effect on existing deployments. Only deployments that explicitly require this
+functionality should turn it on by setting the relevant resource section at
+pod-level in the Pod specification.
+
+### Risks and Mitigations
+
+1. Since Pod Level Resource Specifications is an **opt-in** feature, merging the
+   feature related changes won't impact existing workloads. Moreover, the feature
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+
+2. While this feature doesn't alter the existing cgroups structure, it does
+   change how pod-level cgroup values are determined. Currently, Kubernetes
+   derived these values from the container-level cgroup settings. However, with
+   Pod Level Resource Specifications enabled, pod-level cgroup settings will be
+   directly set based on the values specified in the pod's resource spec stanza.
+
+   This means that any workloads or tools that depend on reading or interpreting
+   container cgroup values might observe different behavior if they start using
+   pod level resource spec without specifying container level settings. This
+   aspect of the feature will be clearly highlighted in the feature documentation
+   to make users aware of this change and the potential impact on their workloads.
+   This way, they can adjust the pod level and container level settings
+   accordingly as needed.
+
+## Design Details
+
+### Components/Features changes
+
+#### Cgroup Structure Remains unchanged
+
+This KEP introduces pod-level resource specifications but does not alter the
+fundamental cgroup structure. Currently, pod level cgroups are derived from the
+resource requests and limits defined at the container level in the PodSpec. With
+this KEP, the PodSpec is extended to include pod-level resource requests and
+limits. These values will be directly used to configure the pod-level cgroup,
+providing a more direct and explicit way to manage the overall resource
+consumption of the pod.
+
+#### PodSpec API changes
+
+New field in `PodSpec`
+
+```
+type PodSpec struct {
+...
+  // Compute resource requirements.
+  // +optional
+  Resources ResourceRequirements
+}
+```
+
+#### PodSpec Validation Rules
+
+##### Proposed Validation Rules
+* **Symmetric Semantic**: Just as container-level requests default to
+  container-level limits when container-level requests are unspecified, the
+  proposal extends this logic to the pod level: if the pod-level request is
+  not specified, it defaults to the pod level limit.
```

Comment:
This makes sense when container requests aren't specified, but I think it could be surprising in the case when they are. Consider this case:

- I have a Burstable pod with a couple containers that specify requests << limits
- Although each container can burst up, I want to cap the simultaneous burst limit, so I add a pod-level limit
- Since I didn't set a pod-level request, the pod request is defaulted to the limit, and the pod is now guaranteed

This seems like it would be a relatively common use-case, and to support it with the requests=limits defaulting, I would need to manually total the container requests (which can be complicated with init containers & sidecars).

Reply:
My expectation for how this would work is:
- If container requests are set, then the pod request is the sum of the container requests (meaning that adding a pod limit to an existing pod wouldn't change the requests)
    - If `sum(container requests) > pod limits`, then fail validation
- If container requests are not set, but pod limit is set, then pod request == pod limit

Reply:
@tallclair IMO it gets complicated in the mix containers scenario: one container has request set, the second one doesn't. What do we do in this case? Should we derive the pod level request from container level requests only when request is set for all containers? 

We could derive pod level requests from container level requests if it is set for all containers only. But it does lead to different behavior in different scenarios which requires more thinking about how the values are derived. We rather wanted to make this simple to use.

If, for a user, it is complicated to derive pod level request from container level requests (although they know the requests for all their containers), one could argue that deriving pod level limit is also challenging but the user did specify the limit at pod level. So they should be able to estimate the minimum resource guarantees they would need for the overall pod level?

Reply:
Trying to think about this again with the following example:

Consider a pod with Pod Limit = 200 Gi; init container request = 50Gi, regular container with no request and limit. 
KEP proposed approach:
Derived Pod Request = 200 Gi; Derived init container limit = 200 Gi, Derived regular container request = min, Derived regular container limit = 200 Gi.

We will document that this is how the defaults would be applied. If you, as a user, want a different memory guarantee at pod level then please specify it. 

@tallclair could you please help me understand what the derived values will look like with the approach you are suggesting?